File: src/main/java/org/micoli/php/configuration/documentation/ClassPropertiesDocumentationGenerator.java
package org.micoli.php.configuration.documentation;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import io.swagger.v3.oas.annotations.media.Schema;
import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.util.*;
import java.util.stream.Collectors;
import org.jetbrains.annotations.Nullable;

public final class ClassPropertiesDocumentationGenerator {

    public record PropertyInfo(
            String dotNotationPath,
            String name,
            Class<?> type,
            String description,
            String example,
            String defaultValue) {}

    private final Set<Class<?>> visitedClasses = new HashSet<>();
    private final List<PropertyInfo> properties = new ArrayList<>();

    public List<PropertyInfo> getProperties(Object object, int maxDepth) {
        visitedClasses.clear();
        traverseClass(object, "", maxDepth);
        return properties;
    }

    private void traverseClass(Object object, String currentPath, int maxDepth) {
        if (maxDepth == 0) {
            return;
        }
        if (object == null
                || visitedClasses.contains(object.getClass())
                || isPrimitiveOrWrapper(object.getClass())
                || object.getClass() == String.class
                || object.getClass().getName().startsWith("java.")) {
            return;
        }

        visitedClasses.add(object.getClass());

        for (Field field : object.getClass().getFields()) {
            if (!Modifier.isPublic(field.getModifiers())) {
                continue;
            }
            if (Modifier.isStatic(field.getModifiers())) {
                continue;
            }
            Object fieldValue;
            try {
                fieldValue = field.get(object);
            } catch (IllegalAccessException e) {
                throw new RuntimeException(e);
            }
            Class<?> fieldTypeCandidate = field.getType();
            Map<String, Class<?>> fieldTypes;
            Schema fieldSchema = field.getAnnotation(Schema.class);
            JsonSubTypes subtypes = fieldTypeCandidate.getAnnotation(JsonSubTypes.class);
            JsonTypeInfo jsonTypeInfo = fieldTypeCandidate.getAnnotation(JsonTypeInfo.class);
            if (subtypes != null) {
                fieldTypes = Arrays.stream(subtypes.value())
                        .collect(Collectors.toMap(JsonSubTypes.Type::name, JsonSubTypes.Type::value));
            } else {
                fieldTypes = Map.of(fieldTypeCandidate.getSimpleName(), fieldTypeCandidate);
            }
            for (Map.Entry<String, Class<?>> subType : fieldTypes.entrySet()) {

                Class<?> fieldType = subType.getValue();
                boolean isMultiple = (fieldValue != null
                        && (fieldValue.getClass().isArray()
                                || Collection.class.isAssignableFrom(fieldType)
                                || Map.class.isAssignableFrom(fieldType)));

                String fieldName = field.getName() + (isMultiple ? "[]" : "");
                String fieldPath = currentPath.isEmpty() ? fieldName : currentPath + "." + fieldName;
                if (jsonTypeInfo != null) {
                    fieldPath = fieldPath + "(" + jsonTypeInfo.property() + "=" + subType.getKey() + ")";
                }

                properties.add(new PropertyInfo(
                        fieldPath,
                        fieldName,
                        fieldType,
                        getDescription(fieldSchema),
                        getExample(fieldSchema),
                        getDefaultValue(isMultiple, fieldValue)));

                if (jsonTypeInfo != null) {
                    properties.add(new PropertyInfo(
                            fieldPath + "." + jsonTypeInfo.property(),
                            fieldName,
                            fieldType,
                            subType.getKey(),
                            subType.getKey(),
                            subType.getKey()));
                }

                if (fieldValue != null && fieldValue.getClass().isArray()) {
                    Object[] values = (Object[]) fieldValue;
                    if (values.length > 0) {
                        traverseClass(values[0], fieldPath, maxDepth - 1);
                    }
                }

                if (Collection.class.isAssignableFrom(fieldType)) {
                    assert fieldValue instanceof Collection<?>;
                    Collection<?> valueCollection = (Collection<?>) fieldValue;
                    if (!valueCollection.isEmpty()) {
                        traverseClass(valueCollection.iterator().next(), fieldPath, maxDepth);
                    }
                }

                if (Map.class.isAssignableFrom(fieldType)) {
                    throw new UnsupportedOperationException("Maps are not yet supported");
                }

                if (!isPrimitiveOrWrapper(fieldType) && fieldType != String.class) {
                    traverseClass(fieldValue, fieldPath, maxDepth - 1);
                }
            }
        }
    }

    private @Nullable String getDefaultValue(boolean isMultiple, Object fieldValue) {
        if (isMultiple) {
            return null;
        }

        if (fieldValue instanceof String stringValue) {
            return stringValue;
        }
        if (fieldValue != null && isPrimitiveOrWrapper(fieldValue.getClass())) {
            return fieldValue.toString();
        }
        return null;
    }

    private String getExample(Schema fieldSchema) {
        if (fieldSchema == null || fieldSchema.example().isEmpty()) {
            return null;
        }
        return fieldSchema.example();
    }

    private String getDescription(Schema fieldSchema) {
        if (fieldSchema == null || fieldSchema.description().isEmpty()) {
            return null;
        }
        return fieldSchema.description();
    }

    private boolean isPrimitiveOrWrapper(Class<?> clazz) {
        return clazz.isPrimitive()
                || clazz == Boolean.class
                || clazz == Character.class
                || clazz == Byte.class
                || clazz == Short.class
                || clazz == Integer.class
                || clazz == Long.class
                || clazz == Float.class
                || clazz == Double.class;
    }
}

File: src/main/kotlin/org/micoli/php/configuration/documentation/ClassPropertiesDocumentationGenerator.kt
package org.micoli.php.configuration.documentation

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo
import io.swagger.v3.oas.annotations.media.Schema
import java.lang.reflect.Field
import java.lang.reflect.Modifier

class ClassPropertiesDocumentationGenerator {
    @JvmRecord
    data class PropertyInfo(
        val dotNotationPath: String,
        val name: String,
        val type: Class<*>,
        val description: String?,
        val example: String?,
        val defaultValue: String?,
    )

    private val visitedClasses: MutableSet<Class<*>> = HashSet()
    private val properties: MutableList<PropertyInfo> = ArrayList<PropertyInfo>()

    fun getProperties(clazz: Any?, maxDepth: Int): MutableList<PropertyInfo> {
        visitedClasses.clear()
        traverseClass(clazz, "", maxDepth)
        return properties
    }

    private fun traverseClass(clazz: Any?, currentPath: String, maxDepth: Int) {
        if (maxDepth == 0) {
            return
        }
        if (clazz == null ||
            visitedClasses.contains(clazz.javaClass) ||
            isPrimitiveOrWrapper(clazz.javaClass) ||
            clazz.javaClass == String::class.java ||
            clazz.javaClass.getName().startsWith("java.")) {
            return
        }

        visitedClasses.add(clazz.javaClass)

        val fields: MutableSet<Field> = HashSet()
        var currentClass: Class<*>? = clazz.javaClass
        // System.out.println(currentPath)
        while (currentClass != null) {
            fields.addAll(listOf(*currentClass.getDeclaredFields()))
            fields.addAll(listOf(*currentClass.getFields()))
            currentClass = currentClass.superclass
        }
        for (field in fields) {
            field.setAccessible(true)

            if (Modifier.isStatic(field.modifiers)) {
                continue
            }
            val fieldValue: Any?
            try {
                fieldValue = field.get(clazz)
            } catch (e: IllegalAccessException) {
                throw RuntimeException(e)
            }
            val fieldTypeCandidate = field.type
            val fieldTypes: MutableMap<String, Class<*>> = HashMap()
            val fieldSchema = field.getAnnotation(Schema::class.java)
            val subtypes = fieldTypeCandidate.getAnnotation(JsonSubTypes::class.java)
            val jsonTypeInfo = fieldTypeCandidate.getAnnotation(JsonTypeInfo::class.java)
            if (subtypes != null) {
                for (value in subtypes.value) {
                    fieldTypes[value.name] = value.value.java
                }
            } else {
                fieldTypes[fieldTypeCandidate.getSimpleName()] = fieldTypeCandidate.javaClass
            }
            for (subType in fieldTypes.entries) {
                val fieldType: Class<*> = subType.value
                val isMultiple =
                    (fieldValue != null &&
                        (fieldValue.javaClass.isArray ||
                            MutableCollection::class.java.isAssignableFrom(fieldType) ||
                            MutableMap::class.java.isAssignableFrom(fieldType)))

                val fieldName = field.name + (if (isMultiple) "[]" else "")
                var fieldPath = if (currentPath.isEmpty()) fieldName else "$currentPath.$fieldName"
                if (jsonTypeInfo != null) {
                    fieldPath = fieldPath + "(" + jsonTypeInfo.property + "=" + subType.key + ")"
                }

                properties.add(
                    PropertyInfo(
                        fieldPath,
                        fieldName,
                        fieldType,
                        getDescription(fieldSchema),
                        getExample(fieldSchema),
                        getDefaultValue(isMultiple, fieldValue),
                    ))

                if (jsonTypeInfo != null) {
                    properties.add(
                        PropertyInfo(
                            fieldPath + "." + jsonTypeInfo.property,
                            fieldName,
                            fieldType,
                            subType.key,
                            subType.key,
                            subType.key,
                        ))
                }

                if (fieldValue != null && fieldValue.javaClass.isArray) {
                    val values = fieldValue as Array<*>
                    if (values.isNotEmpty()) {
                        traverseClass(values[0], fieldPath, maxDepth - 1)
                    }
                }

                if (MutableCollection::class.java.isAssignableFrom(fieldType)) {
                    assert(fieldValue is MutableCollection<*>)
                    val valueCollection = fieldValue as MutableCollection<*>
                    if (!valueCollection.isEmpty()) {
                        traverseClass(valueCollection.iterator().next(), fieldPath, maxDepth)
                    }
                }

                if (MutableMap::class.java.isAssignableFrom(fieldType)) {
                    throw UnsupportedOperationException("Maps are not yet supported")
                }

                if (!isPrimitiveOrWrapper(fieldType) && fieldType != String::class.java) {
                    traverseClass(fieldValue, fieldPath, maxDepth - 1)
                }
            }
        }
    }

    private fun getDefaultValue(isMultiple: Boolean, fieldValue: Any?): String? {
        if (isMultiple) {
            return null
        }

        if (fieldValue is String) {
            return fieldValue
        }
        if (fieldValue != null && isPrimitiveOrWrapper(fieldValue.javaClass)) {
            return fieldValue.toString()
        }
        return null
    }

    private fun getExample(fieldSchema: Schema?): String? {
        if (fieldSchema == null || fieldSchema.example.isEmpty()) {
            return null
        }
        return fieldSchema.example
    }

    private fun getDescription(fieldSchema: Schema?): String? {
        if (fieldSchema == null || fieldSchema.description.isEmpty()) {
            return null
        }
        return fieldSchema.description
    }

    private fun isPrimitiveOrWrapper(clazz: Class<*>): Boolean =
        clazz.isPrimitive ||
            clazz == Boolean::class.javaObjectType ||
            clazz == Char::class.javaObjectType ||
            clazz == Byte::class.javaObjectType ||
            clazz == Short::class.javaObjectType ||
            clazz == Int::class.javaObjectType ||
            clazz == Long::class.javaObjectType ||
            clazz == Float::class.javaObjectType ||
            clazz == Double::class.javaObjectType
}
